### DISCUSSION TITLE
Converting tokio's channels to Streamsink

### DISCUSSION THREAD

**Discussion by acul009** [fzyzcjy/flutter_rust_bridge#2234] [Q&A]
Hi,

I've been using using flutter_rust_bridge for a while now and it's been an absolute blast. I can write my rust code normally and the just create a beautiful ui in flutter without any major headaches.

But I've hit a bit of a roadblock: my library returns different kinds of tokio channel receivers - which I would like to listen to in dart.
While there is the Streamsink, I didn't yet figure out how to easily convert those to a Streamsink, so you have any tips?

An example would be the `tokio::sync::broadcast::Receiver<T>` type which has this function:
```rust
    pub async fn recv(&mut self) -> Result<T, RecvError>
```

I was thinking I could do something akin to this?

```rust
use anyhow::anyhow;
pub use anyhow::Result;

use crate::frb_generated::{SseEncode, StreamSink};

pub trait ToStreamSink {
    type T: Clone;

    async fn streamsink(&mut self, sink: StreamSink<Self::T>) -> Result<()>;
}

impl<T> ToStreamSink for tokio::sync::broadcast::Receiver<T>
where
    T: Clone + SseEncode,
{
    type T = T;

    async fn streamsink(&mut self, sink: StreamSink<Self::T>) -> Result<()> {
        while let Ok(msg) = self.recv().await {
            if let Err(err) = sink.add(msg) {
                return Err(anyhow!(err));
            }
        }

        Ok(())
    }
}
```

Do you have any tips?

PS: @fzyzcjy Do you accept donations? This project is really awesome and I'd love to support it.

**Comment by fzyzcjy** [comment]
Yes, your sample code looks reasonable. Just one thing: frb does not yet support generics. So maybe use Rust's macros (e.g. https://doc.rust-lang.org/rust-by-example/macros.html) to get rid of generics.

> I've been using using flutter_rust_bridge for a while now and it's been an absolute blast. I can write my rust code normally and the just create a beautiful ui in flutter without any major headaches.
> Do you accept donations? This project is really awesome and I'd love to support it.

Thank you for you kindness and happy to hear it works great for you! I do open source primarily because I have been helped by so many open source libraries thus I want to contribute and help other developers, and also because I enjoy coding.

**Reply by acul009** [reply]
Do you think I could manage to add that to the code generator?

i could inject the trait definition into frb_generated.rs, but is it possible to detect when a type implements that trait?

**Reply by fzyzcjy** [reply]
> Do you think I could manage to add that to the code generator?

Generics is not super hard to implement, since we already have all the needed ingredients, and thus we roughly only need to consider `Apple<Orange>` as if it is a brand new `AppleOrange`. However, do not expect it to be implemented as quickly as a dozen lines of code!

As for the macro workaround: That may not be very verbose. For example, the code will look like:

1. use macro_rules to define the macro (slightly longer than your current code using generics because we need to write down macro boilerplate, but < a dozen line I guess)
2. for each type you use, say `Receiver<Apple>, Receiver<Orange>, ...`, call that macro once.

I would suggest using macro_rules that understands repetitions. Then it can be even easier to call, like: `your_macro_name!(Apple, Orange, Another, YetAnother);`

**Reply by acul009** [reply]
So I've been pondering about this for a while now.

No matter what I try, the resulting code is just cumbersome to use.

I think I'll wait for #2189 and maybe try the same for the `futures::Stream` trait.

**Reply by fzyzcjy** [reply]
No worries and take your time!
